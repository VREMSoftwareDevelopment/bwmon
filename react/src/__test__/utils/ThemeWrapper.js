import React from 'react';
import { ThemeProvider, StyledEngineProvider } from '@mui/material/styles';
import { create } from 'react-test-renderer';
import theme from './../../theme';

export const themeWrapper = (component) =>
    create(
        <StyledEngineProvider injectFirst>
            <ThemeProvider theme={theme}>{component}</ThemeProvider>
        </StyledEngineProvider>
    );
