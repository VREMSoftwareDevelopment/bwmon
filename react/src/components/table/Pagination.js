/*
 *      Copyright (C) 2010 - 2023 VREM Software Development <VREMSoftwareDevelopment@gmail.com>
 *
 *      Licensed under the Apache License, Version 2.0 (the "License");
 *      you may not use this file except in compliance with the License.
 *      You may obtain a copy of the License at
 *
 *           http: //www.apache.org/licenses/LICENSE-2.0
 *
 *      Unless required by applicable law or agreed to in writing, software
 *      distributed under the License is distributed on an "AS IS" BASIS,
 *      WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *      See the License for the specific language governing permissions and
 *      limitations under the License.
 *
 * Bandwidth Monitor
 */

import React from 'react';
import { makeStyles, useTheme } from '@mui/styles';
import { TablePagination, IconButton } from '@mui/material';
import { FirstPage, KeyboardArrowLeft, KeyboardArrowRight, LastPage } from '@mui/icons-material';

const useStyles = makeStyles((theme) => ({
    root: {
        flexShrink: 0,
    },
}));

const lastPage = (count, rowsPerPage) => Math.ceil(count / rowsPerPage) - 1;

const isRTL = (theme) => {
    return theme.direction === 'rtl';
};

const isFirstPage = (page) => {
    return page === 0;
};

const isLastPage = (page, count, rowsPerPage) => {
    return page >= lastPage(count, rowsPerPage);
};

const FirstPageAction = ({ page, onPageChange }) => {
    const theme = useTheme();

    const handleFirstPageButtonClick = (event) => {
        onPageChange(event, 0);
    };

    return (
        <IconButton onClick={handleFirstPageButtonClick} disabled={isFirstPage(page)} aria-label="first page">
            {isRTL(theme) ? <LastPage /> : <FirstPage />}
        </IconButton>
    );
};

const LastPageAction = ({ count, page, rowsPerPage, onPageChange }) => {
    const theme = useTheme();

    const handleLastPageButtonClick = (event) => {
        onPageChange(event, Math.max(0, lastPage(count, rowsPerPage)));
    };

    return (
        <IconButton onClick={handleLastPageButtonClick} disabled={isLastPage(page, count, rowsPerPage)} aria-label="last page">
            {isRTL(theme) ? <FirstPage /> : <LastPage />}
        </IconButton>
    );
};

const PreviousPageAction = ({ page, onPageChange }) => {
    const theme = useTheme();

    const handleBackButtonClick = (event) => {
        onPageChange(event, page - 1);
    };

    return (
        <IconButton onClick={handleBackButtonClick} disabled={isFirstPage(page)} aria-label="previous page">
            {isRTL(theme) ? <KeyboardArrowRight /> : <KeyboardArrowLeft />}
        </IconButton>
    );
};

const NextPageAction = ({ count, page, rowsPerPage, onPageChange }) => {
    const theme = useTheme();

    const handleNextButtonClick = (event) => {
        onPageChange(event, page + 1);
    };

    return (
        <IconButton onClick={handleNextButtonClick} disabled={isLastPage(page, count, rowsPerPage)} aria-label="next page">
            {isRTL(theme) ? <KeyboardArrowLeft /> : <KeyboardArrowRight />}
        </IconButton>
    );
};

const Actions = ({ count, page, rowsPerPage, onPageChange }) => {
    const classes = useStyles();

    return (
        <div className={classes.root}>
            <FirstPageAction page={page} onPageChange={onPageChange} />
            <PreviousPageAction page={page} onPageChange={onPageChange} />
            <NextPageAction page={page} onPageChange={onPageChange} count={count} rowsPerPage={rowsPerPage} />
            <LastPageAction page={page} onPageChange={onPageChange} count={count} rowsPerPage={rowsPerPage} />
        </div>
    );
};

const rowsPerPageOptions = (rowsPerPage, count) => {
    let rowsPerPageOptions = [rowsPerPage];
    if (count > rowsPerPage) {
        for (let i = rowsPerPage * 2; i < count; i = i * 2) {
            rowsPerPageOptions.push(i);
        }
        rowsPerPageOptions.push(count);
    }
    return rowsPerPageOptions;
};

const Pagination = (props) => {
    const options = rowsPerPageOptions(Math.min(props.minimum, props.count), props.count);
    const selectProps = {
        inputProps: { 'aria-label': 'rows per page' },
        native: true,
        id: 'select-rows-per-page-id',
        labelId: 'select-label-rows-per-page-id',
    };

    return (
        <TablePagination
            {...props}
            labelRowsPerPage=""
            SelectProps={selectProps}
            rowsPerPageOptions={options}
            ActionsComponent={Actions}
        />
    );
};

export { Pagination, rowsPerPageOptions };

export default Pagination;
