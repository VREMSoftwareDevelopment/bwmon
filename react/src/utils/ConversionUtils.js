/*
 *      Copyright (C) 2010 - 2020 VREM Software Development <VREMSoftwareDevelopment@gmail.com>
 *
 *      Licensed under the Apache License, Version 2.0 (the "License");
 *      you may not use this file except in compliance with the License.
 *      You may obtain a copy of the License at
 *
 *           http: //www.apache.org/licenses/LICENSE-2.0
 *
 *      Unless required by applicable law or agreed to in writing, software
 *      distributed under the License is distributed on an "AS IS" BASIS,
 *      WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *      See the License for the specific language governing permissions and
 *      limitations under the License.
 *
 * Bandwidth Monitor
 */

import { DateTime } from 'luxon';

export const timeToDate = (time) => DateTime.fromSeconds(Number(time)).toFormat('MMM dd, yyyy HH:mm');
export const toMonth = (month) => DateTime.local(2012, month).toFormat('MMMM');
export const usageInGBytes = (value) => (Number(value) / 1000000).toFixed(3);
export const toPercent = (value) => Number(value).toFixed(1) + '%';
export const toIPv4 = (value) => (value >>> 24) + '.' + ((value >> 16) & 255) + '.' + ((value >> 8) & 255) + '.' + (value & 255);
export const fromIPv4 = (value) => value.split('.').reduce((value, octet) => (value << 8) + parseInt(octet, 10), 0) >>> 0;
