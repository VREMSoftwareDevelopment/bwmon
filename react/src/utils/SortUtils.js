/*
 *      Copyright (C) 2010 - 2024 VREM Software Development <VREMSoftwareDevelopment@gmail.com>
 *
 *      Licensed under the Apache License, Version 2.0 (the "License");
 *      you may not use this file except in compliance with the License.
 *      You may obtain a copy of the License at
 *
 *           http: //www.apache.org/licenses/LICENSE-2.0
 *
 *      Unless required by applicable law or agreed to in writing, software
 *      distributed under the License is distributed on an "AS IS" BASIS,
 *      WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *      See the License for the specific language governing permissions and
 *      limitations under the License.
 *
 * Bandwidth Monitor
 */

const isAscending = (orderBy, property, ascending) => (orderBy === property ? !ascending : false);

const comparator = (isAscending, orderBy) => {
    const ascending = (a, b, orderBy) => {
        if (a[orderBy] < b[orderBy]) return -1;
        if (a[orderBy] > b[orderBy]) return 1;
        return 0;
    };
    const descending = (a, b, orderBy) => {
        if (a[orderBy] > b[orderBy]) return -1;
        if (a[orderBy] < b[orderBy]) return 1;
        return 0;
    };
    return isAscending ? (a, b) => ascending(a, b, orderBy) : (a, b) => descending(a, b, orderBy);
};

const sort = (values, comparator) => [...values].sort(comparator);

export { comparator, isAscending, sort };
