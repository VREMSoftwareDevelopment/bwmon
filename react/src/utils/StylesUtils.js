import { makeStyles } from '@mui/styles';

const defaultStyles = makeStyles({});

const footerStyles = makeStyles((theme) => ({
    root: {
        flexGrow: 1,
        margin: theme.spacing(2),
    },
}));

const graphStyles = makeStyles((theme) => ({
    root: {
        flexGrow: 1,
        margin: theme.spacing(2),
        marginTop: 6,
    },
}));

const headerStyles = makeStyles({
    title: {
        flexGrow: 1,
    },
});

const messageStyles = makeStyles((theme) => ({
    root: {
        width: '100%',
        '& > * + *': {
            marginTop: theme.spacing(2),
        },
    },
}));

const paginationStyles = makeStyles((_theme) => ({
    root: {
        flexShrink: 0,
    },
}));

export { defaultStyles, footerStyles, graphStyles, headerStyles, messageStyles, paginationStyles };
