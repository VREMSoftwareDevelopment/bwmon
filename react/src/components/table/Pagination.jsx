/*
 *      Copyright (C) 2010 - 2024 VREM Software Development <VREMSoftwareDevelopment@gmail.com>
 *
 *      Licensed under the Apache License, Version 2.0 (the "License");
 *      you may not use this file except in compliance with the License.
 *      You may obtain a copy of the License at
 *
 *           http: //www.apache.org/licenses/LICENSE-2.0
 *
 *      Unless required by applicable law or agreed to in writing, software
 *      distributed under the License is distributed on an "AS IS" BASIS,
 *      WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *      See the License for the specific language governing permissions and
 *      limitations under the License.
 *
 * Bandwidth Monitor
 */

import React from 'react';
import PropTypes from 'prop-types';
import { useTheme } from '@mui/styles';
import { TablePagination, IconButton } from '@mui/material';
import { FirstPage, KeyboardArrowLeft, KeyboardArrowRight, LastPage } from '@mui/icons-material';
import { paginationStyles } from '../../utils/StylesUtils';

const lastPage = (count, rowsPerPage) => Math.ceil(count / rowsPerPage) - 1;

const isRTL = (theme) => {
    return theme.direction === 'rtl';
};

const isFirstPage = (page) => {
    return page === 0;
};

const isLastPage = (page, count, rowsPerPage) => {
    return page >= lastPage(count, rowsPerPage);
};

const FirstPageAction = ({ page, onPageChange }) => {
    const theme = useTheme();

    const handleFirstPageButtonClick = (event) => {
        onPageChange(event, 0);
    };

    return (
        <IconButton onClick={handleFirstPageButtonClick} disabled={isFirstPage(page)} aria-label="first page">
            {isRTL(theme) ? <LastPage /> : <FirstPage />}
        </IconButton>
    );
};

// Stryker disable next-line all
FirstPageAction.propTypes = { page: PropTypes.number.isRequired, onPageChange: PropTypes.func.isRequired };

const LastPageAction = ({ count, page, rowsPerPage, onPageChange }) => {
    const theme = useTheme();

    const handleLastPageButtonClick = (event) => {
        onPageChange(event, Math.max(0, lastPage(count, rowsPerPage)));
    };

    return (
        <IconButton onClick={handleLastPageButtonClick} disabled={isLastPage(page, count, rowsPerPage)} aria-label="last page">
            {isRTL(theme) ? <FirstPage /> : <LastPage />}
        </IconButton>
    );
};

// Stryker disable all
LastPageAction.propTypes = {
    count: PropTypes.number.isRequired,
    page: PropTypes.number.isRequired,
    rowsPerPage: PropTypes.number.isRequired,
    onPageChange: PropTypes.func.isRequired,
};
// Stryker restore all

const PreviousPageAction = ({ page, onPageChange }) => {
    const theme = useTheme();

    const handleBackButtonClick = (event) => {
        onPageChange(event, page - 1);
    };

    return (
        <IconButton onClick={handleBackButtonClick} disabled={isFirstPage(page)} aria-label="previous page">
            {isRTL(theme) ? <KeyboardArrowRight /> : <KeyboardArrowLeft />}
        </IconButton>
    );
};

// Stryker disable next-line all
PreviousPageAction.propTypes = { page: PropTypes.number.isRequired, onPageChange: PropTypes.func.isRequired };

const NextPageAction = ({ count, page, rowsPerPage, onPageChange }) => {
    const theme = useTheme();

    const handleNextButtonClick = (event) => {
        onPageChange(event, page + 1);
    };

    return (
        <IconButton onClick={handleNextButtonClick} disabled={isLastPage(page, count, rowsPerPage)} aria-label="next page">
            {isRTL(theme) ? <KeyboardArrowLeft /> : <KeyboardArrowRight />}
        </IconButton>
    );
};

// Stryker disable all
NextPageAction.propTypes = {
    count: PropTypes.number.isRequired,
    page: PropTypes.number.isRequired,
    rowsPerPage: PropTypes.number.isRequired,
    onPageChange: PropTypes.func.isRequired,
};
// Stryker restore all

const Actions = ({ count, page, rowsPerPage, onPageChange }) => {
    const classes = paginationStyles();

    return (
        <div className={classes.root}>
            <FirstPageAction page={page} onPageChange={onPageChange} />
            <PreviousPageAction page={page} onPageChange={onPageChange} />
            <NextPageAction page={page} onPageChange={onPageChange} count={count} rowsPerPage={rowsPerPage} />
            <LastPageAction page={page} onPageChange={onPageChange} count={count} rowsPerPage={rowsPerPage} />
        </div>
    );
};

// Stryker disable all
Actions.propTypes = {
    count: PropTypes.number.isRequired,
    page: PropTypes.number.isRequired,
    rowsPerPage: PropTypes.number.isRequired,
    onPageChange: PropTypes.func.isRequired,
};
// Stryker restore all

const rowsPerPageOptions = (rowsPerPage, count) => {
    let rowsPerPageOptions = [rowsPerPage];
    if (count > rowsPerPage) {
        for (let i = rowsPerPage * 2; i < count; i = i * 2) {
            rowsPerPageOptions.push(i);
        }
        rowsPerPageOptions.push(count);
    }
    return rowsPerPageOptions;
};

const Pagination = (props) => {
    const options = rowsPerPageOptions(Math.min(props.minimum, props.count), props.count);
    const selectProps = {
        inputProps: { 'aria-label': 'rows per page' },
        native: true,
        id: 'select-rows-per-page-id',
        labelId: 'select-label-rows-per-page-id',
    };

    return (
        <TablePagination
            {...props}
            labelRowsPerPage=""
            slotProps={{ select: selectProps }}
            rowsPerPageOptions={options}
            ActionsComponent={Actions}
        />
    );
};

// Stryker disable next-line all
Pagination.propTypes = { count: PropTypes.number.isRequired, minimum: PropTypes.number.isRequired };

export { Pagination, rowsPerPageOptions, FirstPageAction, LastPageAction, PreviousPageAction, NextPageAction };

export default Pagination;
