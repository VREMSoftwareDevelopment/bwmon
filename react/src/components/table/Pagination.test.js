/*
 *      Copyright (C) 2010 - 2024 VREM Software Development <VREMSoftwareDevelopment@gmail.com>
 *
 *      Licensed under the Apache License, Version 2.0 (the "License");
 *      you may not use this file except in compliance with the License.
 *      You may obtain a copy of the License at
 *
 *           http: //www.apache.org/licenses/LICENSE-2.0
 *
 *      Unless required by applicable law or agreed to in writing, software
 *      distributed under the License is distributed on an "AS IS" BASIS,
 *      WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *      See the License for the specific language governing permissions and
 *      limitations under the License.
 *
 * Bandwidth Monitor
 */

import React from 'react';
import { themeWrapper } from '../../__test__/utils/ThemeWrapper';
import { Pagination, rowsPerPageOptions } from './Pagination';

describe('Pagination', () => {
    const handlePageChange = (event, newPage) => {};

    const handleRowsPerPageChange = (event) => {};

    it('renders correctly', () => {
        const tree = themeWrapper(
            <Pagination
                colSpan={7}
                count={20}
                minumum={5}
                rowsPerPage={10}
                page={0}
                onPageChange={handlePageChange}
                onRowsPerPageChange={handleRowsPerPageChange}
            />
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('rowsPerPageOptions when rows per page is the same as count', () => {
        const rowsPerPage = 10;
        const count = 10;
        const expected = [rowsPerPage];
        const actual = rowsPerPageOptions(rowsPerPage, count);
        expect(actual).toEqual(expected);
    });

    it('rowsPerPageOptions when rows per page is the larger than count', () => {
        const rowsPerPage = 10;
        const count = 9;
        const expected = [rowsPerPage];
        const actual = rowsPerPageOptions(rowsPerPage, count);
        expect(actual).toEqual(expected);
    });

    it('rowsPerPageOptions when rows per page is the less than count', () => {
        const rowsPerPage = 1;
        const count = 20;
        const expected = [1, 2, 4, 8, 16, 20];
        const actual = rowsPerPageOptions(rowsPerPage, count);
        expect(actual).toEqual(expected);
    });
});
