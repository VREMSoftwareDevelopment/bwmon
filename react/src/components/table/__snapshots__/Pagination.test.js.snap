// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination renders correctly 1`] = `
<td
  aria-sort={null}
  className="MuiTableCell-root MuiTableCell-sizeMedium MuiTablePagination-root css-9wcvca-MuiTableCell-root-MuiTablePagination-root"
  colSpan={7}
  minumum={5}
>
  <div
    className="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular MuiTablePagination-toolbar css-78c6dr-MuiToolbar-root-MuiTablePagination-toolbar"
  >
    <div
      className="MuiTablePagination-spacer css-1psng7p-MuiTablePagination-spacer"
    />
    <p
      className="MuiTablePagination-displayedRows css-levciy-MuiTablePagination-displayedRows"
    >
      1â€“10 of 20
    </p>
    <div
      className="makeStyles-root-1"
    >
      <button
        aria-label="first page"
        className="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
        disabled={true}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={-1}
        type="button"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="FirstPageIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M18.41 16.59 13.82 12l4.59-4.59L17 6l-6 6 6 6zM6 6h2v12H6z"
          />
        </svg>
      </button>
      <button
        aria-label="previous page"
        className="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
        disabled={true}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={-1}
        type="button"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="KeyboardArrowLeftIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.41 16.59 10.83 12l4.58-4.59L14 6l-6 6 6 6 1.41-1.41z"
          />
        </svg>
      </button>
      <button
        aria-label="next page"
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="KeyboardArrowRightIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M8.59 16.59 13.17 12 8.59 7.41 10 6l6 6-6 6-1.41-1.41z"
          />
        </svg>
      </button>
      <button
        aria-label="last page"
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="LastPageIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M5.59 7.41 10.18 12l-4.59 4.59L7 18l6-6-6-6zM16 6h2v12h-2z"
          />
        </svg>
      </button>
    </div>
  </div>
</td>
`;
